src/agent/agent.c

```diff
@@ -135,11 +135,47 @@ static int agent_disconnected(UNUSED sd_bus_message *message, void *userdata, UN
 
 static int agent_reset_heartbeat_timer(Agent *agent, sd_event_source **event_source);
 
+static int agent_check_controller_liveness(Agent *agent) {
+        uint64_t diff = 0;
+        uint64_t now = 0;
+
+        if (agent->heartbeat_threshold_msec <= 0) {
+                /* checking liveness by heartbeat disabled since configured threshold is <=0 */
+                return 0;
+        }
+
+        now = get_time_micros();
+        if (now == 0) {
+                bc_log_error("Failed to get the time");
+                return 0;
+        }
+
+        if (now < agent->connection_last_seen) {
+                bc_log_error("Clock skew detected");
+                return 0;
+        }
+
+        diff = now - agent->connection_last_seen;
+        if (diff > (uint64_t) agent->heartbeat_threshold_msec * USEC_PER_MSEC) {
+                bc_log_infof("Did not receive heartbeat from controller since '%d'ms. Disconnecting it...",
+                             agent->heartbeat_threshold_msec);
+                agent_disconnected(NULL, agent, NULL);
+                return -1;
+        }
+
+        return 0;
+}
+
 static int agent_heartbeat_timer_callback(sd_event_source *event_source, UNUSED uint64_t usec, void *userdata) {
         Agent *agent = (Agent *) userdata;
 
         int r = 0;
         if (agent->connection_state == AGENT_CONNECTION_STATE_CONNECTED) {
+                r = agent_check_controller_liveness(agent);
+                if (r < 0) {
+                        goto out;
```

Comment:
I'd reverse the condition check here so the `goto` is not needed. 
Also, if the `bool` is used as return value for `agent_check_controller_liveness` it gets more concise, e.g.:
```C
if (agent_check_controller_liveness(agent)){
   // emit signal etc.
}
```

Adding it to the connection state condition check would also get rid of the nesting (at the cost of a more complex if):
```C
if ( (agent->connection_state == AGENT_CONNECTION_STATE_CONNECTED) && agent_check_controller_liveness(agent) ){
   // emit signal etc.
} 
else if(...){ ... }
```

Reply:
done